src/test/util/setup_common.cpp

```diff
@@ -120,18 +120,60 @@ BasicTestingSetup::BasicTestingSetup(const ChainType chainType, const std::vecto
         arguments = Cat(arguments, G_TEST_COMMAND_LINE_ARGUMENTS());
     }
     util::ThreadRename("test");
-    fs::create_directories(m_path_root);
-    m_args.ForceSetArg("-datadir", fs::PathToString(m_path_root));
-    gArgs.ForceSetArg("-datadir", fs::PathToString(m_path_root));
     gArgs.ClearPathCache();
     {
         SetupServerArgs(*m_node.args);
+        m_node.args->AddArg("-testdatadir", strprintf("Custom data directory (default: %s<random_string>)", fs::PathToString(fs::temp_directory_path() / "test_common_" PACKAGE_NAME / "")),
+                            ArgsManager::ALLOW_ANY, OptionsCategory::DEBUG_TEST);
         std::string error;
         if (!m_node.args->ParseParameters(arguments.size(), arguments.data(), error)) {
             m_node.args->ClearArgs();
             throw std::runtime_error{error};
         }
     }
+
+    if (!m_node.args->IsArgSet("-testdatadir")) {
+        // By default, the data directory has a random name
+        const auto rand_str{g_insecure_rand_ctx_temp_path.rand256().ToString()};
+        m_path_root = fs::temp_directory_path() / "test_common_" PACKAGE_NAME / rand_str;
+        fs::create_directories(m_path_root);
+    } else {
+        // Custom data directory
+        m_has_custom_datadir = true;
+        fs::path root_dir{m_node.args->GetPathArg("-testdatadir")};
+        if (root_dir.empty()) {
+            std::cerr << "-testdatadir argument is empty, please specify a path\n";
+            exit(EXIT_FAILURE);
+        }
+        const std::string test_path{G_TEST_GET_FULL_NAME ? G_TEST_GET_FULL_NAME() : ""};
+        const fs::path lockdir{root_dir / "test_temp" / fs::PathFromString(test_path)};
+        m_path_root = lockdir / "datadir";
+
+        // Try to obtain the lock; if unsuccessful don't disturb the existing test.
+        try {
+            fs::create_directories(lockdir);
+        } catch (const std::exception&) {
+        }
+        if (!fs::exists(lockdir)) {
+            std::cerr << "Cannot create test data directory " + fs::PathToString(m_path_root) + '\n';
+            exit(EXIT_FAILURE);
+        }
+        if (util::LockDirectory(lockdir, ".lock", /*probe_only=*/false) != util::LockResult::Success) {
+            std::cerr << "Cannot obtain a lock on test data lock directory " + fs::PathToString(lockdir) + '\n' +
+                             "The test executable is probably already running.\n";
+            exit(EXIT_FAILURE);
+        }
+
+        // Always start with a fresh data directory; this doesn't delete the .lock file.
+        fs::remove_all(m_path_root);
+        fs::create_directories(m_path_root);
```

Comment:
If this doesn't delete the .lock file, then the `fs::create_directories` call isn't needed?

Reply:
It's needed because `fs::remove_all(m_path_root)` removes `m_path_root` (and everything below), so we must re-create `m_path_root` (only the last component of the pathname will need to actually be created). I'll change the `fs::create_directories()` to `TryCreateDirectories()` (even though it shouldn't matter here).